runtime/src/main/java/org/corfudb/runtime/collections/TxBuilder.java

```diff
@@ -42,141 +52,291 @@
         return this;
     }
 
-    private void txBegin() {
-        Transaction.TransactionBuilder transactionBuilder = this.objectsView
-                .TXBuild()
-                .type(TransactionType.OPTIMISTIC);
-        if (timestamp != null) {
-            transactionBuilder.snapshot(new Token(timestamp.getEpoch(), timestamp.getSequence()));
+    /**
+     *************************** READ API *****************************************
+     */
+
+    /**
+     * get the full record from the table given a key.
+     * If this is invoked on a Read-Your-Writes transaction, it will result in starting a corfu transaction
+     * and applying all the updates done so far.
+     *
+     * @param table Table object to retrieve the record from
+     * @param key   Key of the record.
+     * @return CorfuStoreEntry<Key, Value, Metadata> instance.
+     */
+    @Nonnull
+    public <K extends Message, V extends Message, M extends Message>
+    CorfuStoreEntry getRecord(@Nonnull Table<K, V, M> table,
+                              @Nonnull final K key) {
+        if (readYourUncommittedWrites) {
+            txBeginInternal();
+            operations.forEach(Runnable::run); // Apply all pending mutations for reads to see it.
+            operations.clear();
+            CorfuRecord<V, M> record = table.get(key);
+            return new CorfuStoreEntry(key, record.getPayload(), record.getMetadata());
+        }
+        try {
+            txBeginInternal();
+            CorfuRecord<V, M> record = table.get(key);
+            return new CorfuStoreEntry(key, record.getPayload(), record.getMetadata());
+        } finally {
+            txEndInternal();
         }
-        transactionBuilder.build().begin();
     }
 
-    private void txEnd() {
-        this.objectsView.TXEnd();
+    /**
+     * Query by a secondary index.
+     *
+     * @param table Table object.
+     * @param indexName Index name. In case of protobuf-defined secondary index it is the field name.
+     * @param indexKey  Key to query.
+     * @param <K>       Type of Key.
+     * @param <V>       Type of Value.
+     * @param <I>       Type of index/secondary key.
+     * @return Result of the query.
+     */
+    @Nonnull
+    public <K extends Message, V extends Message, M extends Message, I extends Comparable<I>>
+    List<CorfuStoreEntry<K, V, M>> getByIndex(@Nonnull Table<K, V, M> table,
+                                              @Nonnull final String indexName,
+                                              @Nonnull final I indexKey) {
+        if (readYourUncommittedWrites) {
+            txBeginInternal();
+            operations.forEach(Runnable::run); // Apply all pending mutations for reads to see it.
+            operations.clear();
+            return table.getByIndex(indexName, indexKey);
+        }
+        try {
+            txBeginInternal();
+            return table.getByIndex(indexName, indexKey);
+        } finally {
+            txEndInternal();
```

Comment:
this will not get executed if readYourUncommittedWrites is true?.  same with other methods

Reply:
yes, the return method earlier bypasses this section if readYourUncommittedWrites is true